frontend/javascripts/oxalis/shaders/segmentation.glsl.js

```diff
@@ -7,29 +7,149 @@ import { hsvToRgb } from "./utils.glsl";
 export const convertCellIdToRGB: ShaderModule = {
   requirements: [hsvToRgb, getRgbaAtIndex],
   code: `
+
+    /*
+      Antialiased step function.
+      Parameter x must not be discontinuous
+
+      See: https://www.shadertoy.com/view/wtjGzt
+    */
+    float aa_step(float x) {
+        float w = fwidth(x);    // pixel width
+        return smoothstep(.7, -.7, (abs(fract(x - .25) - .5) - .25) / w);
+    }
+
+    /*
+      getElementOfPermutation produces a poor-man's permutation of the numbers
+      [1, ..., sequenceLength] and returns the index-th element.
+      The "permutation" is generated using primitive roots.
+      Example:
+        When calling
+          getElementOfPermutation(3, 7, 3)
+        an implicit sequence of
+          7, 2, 6, 4, 5, 1, 3
+        is accessed at index 3 to return a value.
+        Thus, 4 is returned.
+
+      Additional explanation:
+      3 is a primitive root modulo 7. This fact can be used to generate
+      the following pseudo-random sequence by calculating
+      primitiveRoot**index % sequenceLength:
+      3, 2, 6, 4, 5, 1, 3
+      (see https://en.wikipedia.org/wiki/Primitive_root_modulo_n for a more
+      in-depth explanation).
+      Since the above sequence contains 7 elements (as requested), but exactly *one*
+      collision (the first and last elements will always be the same), we swap the
+      first element with 7.
+
+      To achieve a diverse combination with little collisions, multiple, dependent
+      usages of getElementOfPermutation should use unique sequenceLength values
+      which are prime. You can check out this super-dirty code to double-check
+      collisions:
+        https://gist.github.com/philippotto/88487cddcff049c2aac70b69041efacf
+
+      Sample primitiveRoots for different prime sequenceLengths are:
+      sequenceLength=13.0, seed=2
+      sequenceLength=17.0, seed=3
+      sequenceLength=19.0, seed=2
+      More primitive roots for different prime values can be looked up online.
+    */
+    float getElementOfPermutation(float index, float sequenceLength, float primitiveRoot) {
+      float oneBasedIndex = mod(index, sequenceLength) + 1.0;
+      // If oneBasedIndex == 1, isNotFirstElement == 0:
+      float isNotFirstElement = step(1.5, oneBasedIndex);
+
+      float sequenceValue = mod(pow(primitiveRoot, oneBasedIndex), sequenceLength);
+
+      return
+        // Only use sequenceLength if request element is not the first of the sequence
```

Comment:
```suggestion
        // Only use sequenceLength if request element is the first of the sequence
```

Reply:
However, I feel like this would be easier to understand without the not in `isNotFirstElement`.
Does `float isFirstElement = float(oneBasedIndex == 1.0);` work? (from https://stackoverflow.com/questions/13674637/glsl-cast-bool-to-float)